### ISSUE TITLE
Roadmap: Polymer Carbon vis-a-vis Angular 2

### ISSUE THREAD

**Issue by pdemling** [Polymer/project#1]
The recent introduction of the Carbon elements is intriguing and exciting, but it raises the question of how such components could / should work within a larger component-driven app framework, like Google's much anticipated/heralded Angular 2.  Until now, Polymer's natural fit within an Angular 2 world was the front-facing UI layer; it's unclear where Carbon's "application rails" would/should end, and where Angular 2's core services would begin, in the most efficient / complementary manner.

I realize the easy answer is "use whichever one you like", but as someone who loves the direction of both Polymer and Angular 2, I want to integrate both in the best possible way, and so hearing words from either team on shared direction would be quite helpful.  Thanks for any comments (and please feel free to remove this as an issue, if this is not the best place for this type of question).

**Comment by tjsavage** [comment]
Hey @pdemlin, great question, much appreciated. I hope you'll excuse the slightly long-winded answer :smile:

The Polymer library itself is just a light sugaring layer on top of web components. It lets you build in a component model that is native to the web platform. This is very clearly useful to be able to build re-usable UI components, and the majority of the elements we've built on the Polymer project using the Polymer library have been exactly this - UI elements. As you note, these UI elements can play a fairly clearly-delineated role as a UI layer for application frameworks like Angular 2, and this will hopefully continue to be more and more the case as web component adoption grows across browsers.

But while we've been exploring the power that web components bring to the web platform, coupled with the really powerful features that already exist within the platform, we've found that it's entirely feasible - and indeed can be incredibly elegant - to build complex applications out of just composed web components. 

Frameworks filled this massive gap that has existed for much of the web platform's history - that there was no defined, consistent component model. Frameworks constructed and enforced a mental model for how you could structure your app, and so made it much easier to build complex applications. But components built with one framework were fundamentally incompatible with another framework. Now that the web itself has that component model with web components, we think there is a really nice way to use the whole platform idiomatically _as_ your framework.

That's what we're looking to provide with the carbon elements - a set of web components that enable a well-lit path to using the platform in an idiomatic way, so that you can structure complex apps using the platform itself as your framework. There are a whole slew of benefits to this approach of hewing as close to the platform as possible - much lighter dependencies, better performance, easier debugging using the tools baked into the platform itself, better opportunities for browser performance improvements to be reflected in your application, better reusability of components, easier on-boarding of new developers, no massively risky up-front technology decision when it comes to building a web application, etc. The trick is in thinking about the web platform as an application framework. @kevinpschaaf talked a bit about this in his ["Thinking in Polymer"](https://www.youtube.com/watch?v=ZDjiUmx51y8) talk at the Polymer summit. You can always build entire applications thinking this way - this is exactly what Kevin describes - but the carbon elements will provide the rails to make it easier to structure applications using this model. In this sense, Polymer the library is to the web components component model as the carbon elements will be to using the web platform itself as an application framework: the opinionated rails to make it easier to understand and achieve.

So in short - Angular 2 will provide one way of structuring your application that uses Angular's view of what makes a good application structure.  The carbon elements will provide a different way of structuring your application that more directly uses what the web platform itself provides as its structural underpinning. There will be advantages and disadvantages to each approach. And there may be ways to use some carbon elements to solve certain problems within an Angular 2 context.

Ultimately though, the carbon elements and Angular 2 are both works-in-progress, so it remains to be seen how they may be able to complement each other, or if they'll be largely just two different approaches to structuring applications on the web.

**Comment by pdemling** [comment]
Taylor: wow, thanks for such a detailed reply! To get such a thoughtful response from a team leader speaks volumes about the positive role of community feedback in the project.  There’s a lot to ponder here, and I’m excited to see how it all evolves.

One huge question you raise (which seems like one of those eternal questions of software engineering), is *“How much structure is it necessary to impose, in order build a complex application?”*  Because as you rightly point out, it can certainly be feasible and elegant to build solely with well-designed components, composed and mediated in a clear and easy to understand way.  Therefore, if the Carbon elements can slice off some of the services that more prescriptive “frameworks” would normally provide (e.g. routing, i18n, lazy loading, etc), exposing the functionality in a well-mannered, componentized fashion, then the costs of things like framework learning-curve and lock-in can be avoided.  I.e., why be more prescriptively structured than you need?  That’s entirely sensible.

The flip side of this I see however is “why be less prescriptively structured than you need?”.  Because applications aren’t either “simple” or “complex”, in binary fashion; rather, they exist on a *spectrum* of complexity, and the cost/benefit of adopting a prescriptive structure for any given app can be thought as existing on a sliding scale, to be continually evaluated as the codebase matures (at least, in my experience*).  Therefore, why adopt a “framework,” if it’s only going to hem you in to its way of doing things?  The answer for me is, “when the cost of doing so (in terms of prescription) is outweighed by the organizational benefit provided.”  (and I think this what you were saying above, that there will advantages/disadvantages to the evolving Polymer and Angular approaches).

Much like “no one can be told what the Matrix is” however, I find it’s really hard to get this point across (i.e. that prescriptive structure can have value) – I think you have to experience the pain of a complex codebase firsthand, and feel where the cost/benefit tipping point is, before you then go looking for the kind of design/organizational structure that an opinionated “framework” provides.  This is why I’m hoping that the Angular and Polymer teams say a lot publicly in the future about how/where these two great solutions can complement each other, across the broad spectrum of application complexity (especially now that Polymer has started to dip its toe in the more frameworky waters with Carbon).  I think this would go a long way to minimizing developer confusion, and the “Polymer v Angular”, “which one is best” type of questions.  Thanks!

**[I come from an Apache Flex background, a component-based app framework that’s confronted the same kind of challenges for ~the past 10 years.  And while I realize its reliance on the fading/proprietary Flash Player make it persona non grata today, I find it a useful study for the evolution/use of web components, especially regarding the question of when framework adoption is or isn’t a good idea as a means of managing complexity].*

**Comment by abdonrd** [comment]
Few days ago, Peter Burns show a sneak peak about `carbon-router`. [Reference](https://twitter.com/abdonrd/status/710551868586528768).

Yesterday I see that the `chrome://downloads` has a new `i18n` system. [Reference](https://twitter.com/abdonrd/status/711181939353501696).

@tjsavage there are news about **Carbon elements** release date?

**Comment by tjsavage** [comment]
The Chrome UI pages have a pretty baked-in system for i18n. We will be opening up a router prototypes within the next couple weeks, and an i18n prototype soon after. From there it will be iterate with feedback until we're stable!

**Comment by abdonrd** [comment]
Nice news! :smile:

**Comment by jerryf196** [comment]
Sorry but what are carbon elements?

**Comment by abdonrd** [comment]
@jerryf196 https://github.com/Polymer/project/blob/3e466d845bb1781eb50a87e9f3067470db8a0b67/Roadmap.md#application-rails

**Comment by pdemling** [comment]
On the general question of Polymer/Angular2 relationship/divergence, it's notable that Angular2 Material Design components are now in alpha: https://github.com/angular/material2. This echoes @tjsavage 's comments above, that Polymer and Angular2 are different options for how to structure component-driven applications; with each seemingly offering more and more pieces of the puzzle directly. And both very much steeped in MD (as if google is promoting Material Design through multiple channels intentionally).

**Comment by rictic** [comment]
carbon-route is out! https://github.com/PolymerElements/carbon-route

**Comment by rahulmaha** [comment]
I think polymer needs more examples, I have not come across things like session management (non SPA). SPA of polymer (assuming iron pages ) for now is also not very well demonstrated, i.e. how do you **unload modules**, lazy load example (import HREF just got demonstrated in Mar 2016, one year after polymer 1.0). For me it is tough to create enterprise class application out of polymer alone, I am sure it may be possible. Google as a company is promoting both Angular & Polymer, & if they are expected to work together then no official integration examples are available.
